packages/cpp/ArmoniK.Api.Client/source/channel/ChannelFactory.cpp

```diff
@@ -0,0 +1,165 @@
+  }
+  return endpoint_view.back() == 's' || endpoint_view.back() == 'S';
+}
+
+std::shared_ptr<CertificateProviderInterface> create_certificate_provider(const std::string &rootCertificate,
+                                                                          const std::string &userPublicPem,
+                                                                          const std::string &userPrivatePem) {
+  if (rootCertificate.empty()) {
+    return std::make_shared<StaticDataCertificateProvider>(
+        std::vector<IdentityKeyCertPair>{IdentityKeyCertPair{userPrivatePem, userPublicPem}});
+  } else if (userPrivatePem.empty() || userPublicPem.empty()) {
+    return std::make_shared<StaticDataCertificateProvider>(rootCertificate);
+  } else {
+    return std::make_shared<StaticDataCertificateProvider>(
+        rootCertificate, std::vector<IdentityKeyCertPair>{IdentityKeyCertPair{userPrivatePem, userPublicPem}});
+  }
+}
+
+std::shared_ptr<grpc::Channel> ChannelFactory::create_channel() {
+  auto channel = grpc::CreateCustomChannel(endpoint_, credentials_, common::utils::getChannelArguments(configuration_));
+  logger_.log(common::logger::Level::Debug, "Created new channel ");
+
+  if (channel != nullptr) {
+    if (ShutdownOnFailure(channel)) {
+      logger_.log(common::logger::Level::Debug, "Shutdown unhealthy channel");
+    } else {
+      logger_.log(common::logger::Level::Debug, "Valid channel");
+    }
+  }
+  return channel;
+}
+
+ChannelFactory::ChannelFactory(armonik::api::common::utils::Configuration configuration, common::logger::Logger &logger)
+    : logger_(logger.local()), configuration_(std::move(configuration)) {
+  const auto control_plane = configuration_.get_control_plane();
+  const bool is_https = initialize_protocol_endpoint(control_plane, endpoint_);
+
+  auto root_cert_pem = get_key(control_plane.getCaCertPemPath());
+  auto user_private_pem = get_key(control_plane.getUserKeyPemPath());
+  auto user_public_pem = get_key(control_plane.getUserCertPemPath());
+
+  if (is_https) {
+    if (!user_private_pem.empty() && !user_public_pem.empty()) {
+      if (control_plane.isSslValidation()) {
+        credentials_ = grpc::SslCredentials(grpc::SslCredentialsOptions{
+            std::move(root_cert_pem), std::move(user_private_pem), std::move(user_public_pem)});
+      } else {
+        throw common::exceptions::ArmoniKApiException("mTLS without SSL validation is not supported.");
+      }
+    } else {
+      if (control_plane.isSslValidation()) {
+        credentials_ = grpc::SslCredentials(grpc::SslCredentialsOptions{std::move(root_cert_pem)});
+      } else {
+        TlsChannelCredentialsOptions tls_options;
+        tls_options.set_certificate_provider(
+            create_certificate_provider(root_self_signed, user_public_pem, user_private_pem));
+        tls_options.set_verify_server_certs(control_plane.isSslValidation());
+        credentials_ = TlsCredentials(tls_options);

```

Comment:
I thought you could not disable ssl validation

Reply:
It can not be disabled only in the case where the client certificate and key are specified

Reply:
Ok. That's very specific, but I'm actually glad that's only for this very special case that won't happen in prod, and should not be the norm for devs.